scipy/stats/_stats_py.py

```diff
@@ -3213,25 +3214,28 @@ def gstd(a, axis=0, ddof=1):
       fill_value=999999)
 
     """
-    a = np.asanyarray(a)
-    log = ma.log if isinstance(a, ma.MaskedArray) else np.log
+    if isinstance(a, ma.MaskedArray):
+        xp = np
+        log = ma.log
+        ddof = {'ddof': ddof}
+    else:
+        xp = array_namespace(a)
+        log = xp.log
+        ddof = {'correction': ddof}
+        a = xp.asarray(a)
 
     try:
         with warnings.catch_warnings():
             warnings.simplefilter("error", RuntimeWarning)
-            return np.exp(np.std(log(a), axis=axis, ddof=ddof))
+            return xp.exp(xp.std(log(a), axis=axis, **ddof))
     except RuntimeWarning as w:
-        if np.isinf(a).any():
+        if xp.any(xp.isinf(a)):
             raise ValueError(
                 'Infinite value encountered. The geometric standard deviation '
                 'is defined for strictly positive values only.'
             ) from w
-        a_nan = np.isnan(a)
-        a_nan_any = a_nan.any()
-        # exclude NaN's from negativity check, but
-        # avoid expensive masking for arrays with no NaN
-        if ((a_nan_any and np.less_equal(np.nanmin(a), 0)) or
-                (not a_nan_any and np.less_equal(a, 0).any())):
```

Comment:
I don't understand the reason for tiptoeing around NaNs here. AFAICT, the behavior is the same even if NaNs are present, and no warnings are emitted.